frontend/src/concepts/connectionTypes/NewConnectionField.tsx

```diff
@@ -0,0 +1,185 @@
+import {
+  Connection,
+  ConnectionTypeConfigMapObj,
+  ConnectionTypeDataField,
+  ConnectionTypeFieldType,
+  ConnectionTypeValueType,
+} from './types';
+import {
+  assembleConnectionSecret,
+  filterEnabledConnectionTypes,
+  getDefaultValues,
+  isConnectionTypeDataField,
+} from './utils';
+import ConnectionTypeForm from './ConnectionTypeForm';
+
+export type UseNewConnectionFieldData = {
+  connectionTypeOptions: ConnectionTypeConfigMapObj[];
+  selectedConnectionType?: ConnectionTypeConfigMapObj;
+  setSelectedConnectionType: (ct?: ConnectionTypeConfigMapObj) => void;
+  nameDescData: UseK8sNameDescriptionFieldData['data'];
+  setNameDescData: UseK8sNameDescriptionFieldData['onDataChange'];
+  connectionValues: { [key: string]: ConnectionTypeValueType };
+  setConnectionValues: (values: { [key: string]: ConnectionTypeValueType }) => void;
+  connectionErrors: { [key: string]: string | boolean };
+  setConnectionErrors: (errors: { [key: string]: string | boolean }) => void;
+  isFormValid: boolean;
+  extraConnectionTypeModifiers?: (ct: ConnectionTypeConfigMapObj) => ConnectionTypeConfigMapObj;
+  extraValidation?: (
+    field: ConnectionTypeDataField,
+    error: boolean | string,
+    value?: ConnectionTypeValueType,
+  ) => string | boolean | null;
+  newConnection?: Connection;
+};
+
+export const useNewConnectionField = (
+  projectName: string,
+  connectionTypes: ConnectionTypeConfigMapObj[],
+  initialConnectionType?: ConnectionTypeConfigMapObj,
+  initialConnectionValues?: {
+    [key: string]: ConnectionTypeValueType;
+  },
+  extraConnectionTypeModifiers?: UseNewConnectionFieldData['extraConnectionTypeModifiers'],
+  extraValidation?: UseNewConnectionFieldData['extraValidation'],
+): UseNewConnectionFieldData => {
+  const enabledConnectionTypes = React.useMemo(
+    () =>
+      filterEnabledConnectionTypes(connectionTypes).map(
+        (ct) => extraConnectionTypeModifiers?.(ct) ?? ct,
+      ),
+    [connectionTypes, extraConnectionTypeModifiers],
+  );
+
+  const [tmpSelectedConnectionType, setSelectedConnectionType] =
+    React.useState<ConnectionTypeConfigMapObj>();
+  const selectedConnectionType = React.useMemo(() => {
+    if (tmpSelectedConnectionType === undefined) {
+      if (initialConnectionType) {
+        return initialConnectionType;
+      }
+      if (enabledConnectionTypes.length === 1) {
+        return enabledConnectionTypes[0];
+      }
+      return undefined;
+    }
+    return tmpSelectedConnectionType;
+  }, [tmpSelectedConnectionType, initialConnectionType, enabledConnectionTypes]);
+
+  const { data: nameDescData, onDataChange: setNameDescData } = useK8sNameDescriptionFieldData();
+  const [tmpConnectionValues, setConnectionValues] = React.useState<{
+    [key: string]: ConnectionTypeValueType;
+  }>();
+  const connectionValues = React.useMemo(() => {
+    if (tmpConnectionValues === undefined || Object.entries(tmpConnectionValues).length === 0) {
```

Comment:
```suggestion
    if (tmpConnectionValues === undefined) {
```

I'm not sure how i feel about this `Object.entries(tmpConnectionValues).length === 0)`, ideally it could just use `undefined` for the check, as that is what `undefined` is meant to be used for. But basically all instances of `connectionValues` can't be undefined, so will default to `{}`

Reply:
I think this is a consequence of dealing with TS and it evaluates an empty object as truthy. I'm ok with it unless you have other ideas

Reply:
Yeah so i was looking into making `connectionValues` optional (aka `connectionValues?: ...`) but there are so many downstream changes and i couldn't get all the tests working. 

I'm slightly worried there may be some unintended bugs if there is a connection type with no fields (but there really shouldn't, but idk if that's a safe assumption)

Reply:
ok I see what you mean I can try it out real quick and see. A connection type with no fields seems like an edge case anyway since I can't think of why a customer would do that :thinking:

Reply:
Looks fine to me. If you have an empty connection type it doesn't show up in the connection section since it's missing the required fields to be serving compatible